ca/certificates_test.go

```diff
@@ -492,29 +594,70 @@ func TestGetRemoteSignedCertificateWithPending(t *testing.T) {
 	updates, cancel := state.Watch(tc.MemoryStore.WatchQueue(), api.EventCreateNode{})
 	defer cancel()
 
+	fakeCAServer := newNonSigningCAServer(t, tc)
+
 	completed := make(chan error)
+	defer close(completed)
 	go func() {
 		_, err := ca.GetRemoteSignedCertificate(context.Background(), csr, tc.RootCA.Pool,
 			ca.CertificateRequestConfig{
 				Token:      tc.WorkerToken,
-				ConnBroker: tc.ConnBroker,
+				ConnBroker: fakeCAServer.getConnBroker(),
+				// ensure the RPC call to get state is cancelled after 1 second
+				NodeCertificateStatusRequestTimeout: 500 * time.Second,
 			})
 		completed <- err
 	}()
 
-	event := <-updates
-	node := event.(api.EventCreateNode).Node.Copy()
+	var node *api.Node
+	// wait for a new node to show up
+	for node == nil {
+		select {
+		case event := <-updates: // we want to skip the first node, which is the test CA
+			n := event.(api.EventCreateNode).Node.Copy()
+			if n.Certificate.Status.State == api.IssuanceStatePending {
+				node = n
+			}
+		}
+	}
+	// wait for five seconds and ensure that GetRemoteSignedCertificate has not returned with an error yet
+	select {
+	case <-completed:
+		assert.FailNow(t, "GetRemoteSignedCertificate should wait at least 3 seconds")
+	case <-time.After(3 * time.Second):
+		// good, it's still polling so we can proceed with the test
+	}
 
 	// Directly update the status of the store
 	err = tc.MemoryStore.Update(func(tx store.Tx) error {
 		node.Certificate.Status.State = api.IssuanceStateIssued
-
 		return store.UpdateNode(tx, node)
 	})
 	assert.NoError(t, err)
 
 	// Make sure GetRemoteSignedCertificate didn't return an error
 	assert.NoError(t, <-completed)
+
+	// make sure if we cancel the GetRemoteSignedCertificate, it cancels immediately
+	go func() {
+		ctx, _ := context.WithTimeout(context.Background(), 500*time.Millisecond)
+		_, err := ca.GetRemoteSignedCertificate(ctx, csr, tc.RootCA.Pool,
+			ca.CertificateRequestConfig{
+				Token:      tc.WorkerToken,
+				ConnBroker: fakeCAServer.getConnBroker(),
+				// ensure the RPC call to get state is cancelled after 1 second
+				NodeCertificateStatusRequestTimeout: time.Second,
+			})
+		completed <- err
+	}()
+
+	// wait for five seconds and ensure that GetRemoteSignedCertificate has not returned with an error yet
+	select {
+	case err = <-completed:
+		assert.Equal(t, grpc.Code(err), codes.DeadlineExceeded)
+	case <-time.After(570 * time.Second):
```

Comment:
Sorry, yes it was supposed to be `time.Millisecond`.  I can bump this up to 1 second.